Xamarin.Forms.Platform.Android/CollectionView/RecyclerViewScrollListener.cs

```diff
@@ -0,0 +1,116 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Android.Graphics;
+using Android.Support.V7.Widget;
+
+namespace Xamarin.Forms.Platform.Android.CollectionView
+{
+	public class RecyclerViewScrollListener : RecyclerView.OnScrollListener
+	{
+		bool _disposed;
+		int _horizontallOffset, _verticalOffset;
+		ItemsView _itemsView;
+		ItemsViewAdapter _itemsViewAdapter;
+
+		public RecyclerViewScrollListener(ItemsView itemsView, ItemsViewAdapter itemsViewAdapter)
+		{
+			_itemsView = itemsView;
+			_itemsViewAdapter = itemsViewAdapter;
+		}
+
+		public override void OnScrollStateChanged(RecyclerView recyclerView, int newState)
+		{
+			base.OnScrollStateChanged(recyclerView, newState);
+		}
+
+		public override void OnScrolled(RecyclerView recyclerView, int dx, int dy)
+		{
+			base.OnScrolled(recyclerView, dx, dy);
+
+			// TODO: These offsets will be incorrect upon row size or count change.
+			// They are currently provided in place of LayoutManager's default offset calculation
+			// because it does not report accurate values in the presence of uneven rows.
+			// See https://stackoverflow.com/questions/27507715/android-how-to-get-the-current-x-offset-of-recyclerview
+			_horizontallOffset += dx;
+			_verticalOffset += dy;
+
+			var firstVisibleItemIndex = -1;
+			var lastVisibleItemIndex = -1;
+			var centerItemIndex = -1;
+
+			if (recyclerView.GetLayoutManager() is LinearLayoutManager linearLayoutManager)
+			{
+				firstVisibleItemIndex = linearLayoutManager.FindFirstVisibleItemPosition();
+				lastVisibleItemIndex = linearLayoutManager.FindLastVisibleItemPosition();
+				centerItemIndex = CalculateCenterItemIndex(firstVisibleItemIndex, lastVisibleItemIndex, linearLayoutManager);
+			}
+
+			var itemsViewScrolledEventArgs = new ItemsViewScrolledEventArgs
+			{
+				HorizontalDelta = dx,
+				VerticalDelta = dy,
+				HorizontalOffset = _horizontallOffset,
+				VerticalOffset = _verticalOffset,
+				FirstVisibleItemIndex = firstVisibleItemIndex,
+				CenterItemIndex = centerItemIndex,
+				LastVisibleItemIndex = lastVisibleItemIndex
+			};
+
+			_itemsView.SendScrolled(itemsViewScrolledEventArgs);
+
+			switch (_itemsView.RemainingItemsThreshold)
+			{
+				case -1:
+					return;
+				case 0:
+					if (firstVisibleItemIndex == _itemsViewAdapter.ItemCount - 1)
+						_itemsView.SendRemainingItemsThresholdReached();
+					break;
+				default:
+					if (_itemsViewAdapter.ItemCount - 1 - firstVisibleItemIndex <= _itemsView.RemainingItemsThreshold)
+						_itemsView.SendRemainingItemsThresholdReached();
+					break;
+			}
+		}
+
+		static int CalculateCenterItemIndex(int firstVisibleItemIndex, int lastVisibleItemIndex, LinearLayoutManager linearLayoutManager)
+		{
+			var keyValuePairs = new Dictionary<int, int>();
+			for (var i = firstVisibleItemIndex; i <= lastVisibleItemIndex; i++)
+			{
+				var view = linearLayoutManager.FindViewByPosition(i);
```

Comment:
When a `CollectionView` is reset during the app execution, consider the state of the app changes and it requires a `Clear()` on the `ObservableCollection<T>` that is bound to y our `CollectionView.ItemsSource`. 

In this scenario the `firstVisibleItemIndex` can be less than `0`. Typically the index would be reset to `-1`. The code as provided will throw a `NullReferenceException` at line 84 where it attempts to use the returned `view`.

My solution was to put a check at the top of the function:

```c#
if (firstVisibleItemIndext < 0)
    return -1;
```

@adrianknight89 what are your thoughts on this?

Reply:
We shouldn't expect to see -1 at all. What you're seeing is probably due to the fact that a layout pass has not occurred yet. I'm not sure if this issue needs to be addressed somewhere else. I ran into a similar issue on iOS today and made a similar fix. I'll update the Android side as well. Thanks!

Reply:
No problem, I saw it and wanted to make sure you were aware. Perhaps, we should put some check or validation higher up the call stack so it never happens in the calculation method.